core/lib/types/src/kzg.rs

```diff
@@ -0,0 +1,315 @@
+
+        res[ptr..ptr + 48].copy_from_slice(self.kzg_commitment.as_slice());
+        ptr += 48;
+
+        res[ptr..ptr + 32].copy_from_slice(self.opening_point.as_slice());
+        ptr += 32;
+
+        res[ptr..ptr + 32].copy_from_slice(self.opening_value.as_slice());
+        ptr += 32;
+
+        res[ptr..ptr + 48].copy_from_slice(self.opening_proof.as_slice());
+        ptr += 48;
+
+        res[ptr..ptr + 32].copy_from_slice(self.versioned_hash.as_slice());
+        ptr += 32;
+
+        res[ptr..ptr + 48].copy_from_slice(self.blob_proof.as_slice());
+        ptr += 48;
+
+        assert_eq!(ptr, Self::SERIALIZED_SIZE);
+
+        res
+    }
+
+    /// Construct all the KZG info we need for turning a piece of zksync pubdata into a 4844 blob.
+    /// The information we need is:
+    ///     1. zksync blob <- `pad_right`(pubdata)
+    ///     2. linear hash <- hash(zksync blob)
+    ///     3. 4844 blob <- `zksync_pubdata_into_ethereum_4844_data`(zksync blob)
+    ///     4. 4844 `kzg` commitment <- `blob_to_kzg_commitment`(4844 blob)
+    ///     5. versioned hash <- hash(4844 `kzg` commitment)
+    ///     6. opening point <- keccak(linear hash || versioned hash)[16..]
+    ///     7. opening value, opening proof <- `compute_kzg_proof`(4844)
+    ///     8. blob proof <- `compute_blob_kzg_proof`(blob, 4844 `kzg` commitment)
+    pub fn new(kzg_settings: &KzgSettings, pubdata: Vec<u8>) -> Self {
+        assert!(pubdata.len() <= BYTES_PER_BLOB_ZK_SYNC);
+
+        let mut zksync_blob = [0u8; BYTES_PER_BLOB_ZK_SYNC];
+        zksync_blob[0..pubdata.len()].copy_from_slice(&pubdata);
+
+        let mut hasher = Keccak256::new();
+        hasher.update(zksync_blob);
+        let linear_hash = &hasher.finalize_reset();
+
+        let bytes_4844 = zksync_pubdata_into_ethereum_4844_data(&zksync_blob);
+        let blob = Blob::new(bytes_4844.try_into().unwrap());
+
+        let kzg_commitment = KzgCommitment::blob_to_kzg_commitment(&blob, kzg_settings).unwrap();
+
+        let mut sha256_hasher = Sha256::new();
+        sha256_hasher.update(kzg_commitment.to_bytes().into_inner());
+        let mut versioned_hash_bytes = sha256_hasher.finalize();
+        versioned_hash_bytes[0] = VERSIONED_HASH_VERSION_KZG;
+        let versioned_hash = Bytes32::from_bytes(&versioned_hash_bytes).unwrap();
+
+        hasher.update(linear_hash.as_slice());
+        hasher.update(versioned_hash_bytes);
+
+        let opening_point_bytes = &hasher.finalize_reset();
+        let mut opening_point = [0u8; 32];
+        opening_point[16..].copy_from_slice(&opening_point_bytes[16..]);
+        let opening_point = Bytes32::from_bytes(&opening_point).unwrap();
+
+        let (opening_proof, opening_value) =
+            KzgProof::compute_kzg_proof(&blob, &opening_point, kzg_settings).unwrap();
+
+        let blob_proof =
+            KzgProof::compute_blob_kzg_proof(&blob, &kzg_commitment.to_bytes(), kzg_settings)
+                .unwrap();
+
```

Comment:
I'm ultra-confused by this part. 
could you describe what is going on here? why we don't care about first 16 bytes.

Reply:
this is expected on the crypto side [here](https://github.com/matter-labs/zkevm_circuits/blob/f7c610e2ba15e6e426a6f474b722f095c5921dbe/src/eip_4844/mod.rs#L607) ill add in a comment about it